demos/eigenvalues_QG_basinmodes/qgbasinmodes.py.rst

```diff
@@ -154,13 +154,22 @@ We define the Test Function :math:`\phi` and the Trial Function
 To build the weak formulation of our equation we need to build two PETSc
 matrices in the form of a generalized eigenvalue problem,
 :math:`A\psi = \lambda M\psi`. We impose the boundary conditions on the
-mass matrix :math:`M`, since that is where we used integration by parts. ::
+mass matrix :math:`M`, since that is where we used integration by parts.
+For a more stable computation of the eigenvalues, we also create
+submatrices by removing rows and columns associated with the boundary
+conditions. ::
 
    a =  beta*phi*psi.dx(0)*dx
    m = -inner(grad(psi), grad(phi))*dx - F*psi*phi*dx
    petsc_a = assemble(a).M.handle
    petsc_m = assemble(m, bcs=bc).M.handle
 
+   bcnodes = Vcg.dof_dset.lgmap.apply(bc.nodes).astype(np.int32)
+   ISbd = PETSc.IS().createGeneral(bcnodes, comm=PETSc.COMM_SELF)
+   ISin= ISbd.complement(0, Vcg.dim())
+   A = petsc_a.createSubMatrix(ISin, ISin)
+   M = petsc_m.createSubMatrix(ISin, ISin)
```

Comment:
I think that this needs a bit more explanation. Moreover there are a few nitpicks:

1. `np.int32` should be `pyop2.datatypes.IntType` (Aside, `IntType` and `ScalarType` should be exposed on import of Firedrake I think)?
2. naming convention is rather off from normal Python.
3. I think that the communicator for the IS should be `petsc_a.comm` not `COMM_SELF`.
4. I think the call to complement is wrong if running in parallel (see [ISComplement](https://petsc.org/release/docs/manualpages/IS/ISComplement.html)). I think you want to say something like `(s, e), _ = petsc_a.getSizes()` and use `iset.complement(s, e)`

Reply:
1. changed to `pyop2.datatypes.IntType`. I don't understand what exposing means.
2. suggestions are welcome
3. changed to `petsc_a.comm`
4. changed as suggested